Customer unit test

fatiyapici opened the pull request:
Customer için Create-Delete-Update ve Get Detail testleri yapıldı.

_Update Customer Unit Test kontrol edilecek._

fatiyapici committed:
add create customer unit test

fatiyapici committed:
add delete customer unit test

fatiyapici committed:
add update customer unit test

fatiyapici committed:
add create token unit test

fatiyapici committed:
add get customer detail unit test

bataylan commented on MovieStoreWebApi/Applications/CustomerOperations/Commands/CreateCustomer/CreateCustomerCommand.cs:
```diff
@@ -7,6 +7,8 @@ namespace MovieStoreWebApi.CustomerOperations.Commands.CreateCustomer
     public class CreateCustomerCommand
     {
         public CreateCustomerViewModel Model { get; set; }
+        public const string ExceptionMessage = "Kullanici adi zaten mevcut.";
```
const field üste alalım

fatiyapici replied:
Düzeltildi.

bataylan commented on MovieStoreWebApi/Applications/CustomerOperations/Commands/DeleteCustomer/DeleteCustomerCommand.cs:
```diff
@@ -4,28 +4,40 @@ namespace MovieStoreWebApi.Applications.CustomerOperations.DeleteCustomer
         public int CustomerId { get; set; }
         public string CustomerEmail { get; set; }
         public string CustomerPassword { get; set; }
+        public const string ExceptionMessageFound = "Silinecek musteri bulunamadi.";
+        public const string ExceptionMessageEmail = "Silinecek musteri emaili yanlis.";
+        public const string ExceptionMessagePassword = "Silinecek musteri sifresi yanlis.";
```
const field üste alalım

fatiyapici replied:
Düzeltildi.

bataylan commented on MovieStoreWebApi/Applications/CustomerOperations/Commands/DeleteCustomer/DeleteCustomerCommand.cs:
```diff
@@ -4,28 +4,40 @@ namespace MovieStoreWebApi.Applications.CustomerOperations.DeleteCustomer
     public class DeleteCustomerCommand
     {
+        public DeleteCustomerModel Model { get; set; }
         public int CustomerId { get; set; }
         public string CustomerEmail { get; set; }
         public string CustomerPassword { get; set; }
```
Id yi de modele ekle, CustomerId değil Id. 
Bu fieldları kaldırıp model üzerindeki fieldları ekle.

fatiyapici replied:
Düzeltildi.

bataylan commented on MovieStoreWebApi/Applications/CustomerOperations/Commands/UpdateCustomer/UpdateCustomerCommand.cs:
```diff
@@ -6,21 +6,26 @@ public class UpdateCustomerCommand
     {
         public UpdateCustomerViewModel Model { get; set; }
         public int Id { get; set; }
+        public const string ExceptionMessageFound = "Guncellenecek musteri bulunamadi.";
+        public const string ExceptionMessageEmail = "Guncellenecek musteri maili yanlis.";
+        public const string ExceptionMessagePassword = "Guncellenecek musteri sifresi yanlis.";
```
const field üste alalım

fatiyapici replied:
Düzeltildi.

bataylan commented on MovieStoreWebApi/Applications/CustomerOperations/Queries/GetCustomerDetail/GetCustomerDetailById.cs:
```diff
@@ -7,6 +7,8 @@ namespace MovieStoreWebApi.Applications.CustomerOperations.GetCustomerDetail
     public class GetCustomerDetailById
     {
         public int CustomerId { get; set; }
+        public const string ExceptionMessage = "Müşteri bulunamadi.";
```
const field üste alalım

fatiyapici replied:
Düzeltildi.

bataylan commented on MovieStoreWebApi/Applications/CustomerOperations/Queries/GetCustomerDetail/GetCustomerDetailById.cs:
```diff
@@ -18,12 +20,12 @@ public GetCustomerDetailById(IMovieStoreDbContext context, IMapper mapper)
         {
             var customer = _context.Customers
-            .Include(g => g.CustomerOrders).ThenInclude(g => g.Order).ThenInclude(o=>o.Movie)
+            .Include(g => g.CustomerOrders).ThenInclude(g => g.Order).ThenInclude(o => o.Movie)
             .Include(g => g.FavoriteGenres).ThenInclude(g => g.Genre)
             .SingleOrDefault(x => x.Id == CustomerId);
```
indentation hatalı. altında da bir satır boşluk bırakalım if le.

fatiyapici replied:
Düzeltildi.

bataylan commented on Tests/WebApi.UnitTests/Applications/CustomerOperations/Commands/DeleteCustomer/DeleteCustomerCommandTest.cs:
```diff
@@ -0,0 +1,57 @@
+        [Fact]
+        public void WhenDeletedCustomerIsNotExist_InvalidOperationException_ShouldReturn()
+        {
+            var person = new Person()
+            {
+                Name = "Burak",
+                Surname = "Hassemercioglu"
+            };
+            _context.Persons.Add(person);
+            _context.SaveChanges();
+
+            var customer = new Customer
+            {
+                PersonId = person.Id,
+                Email = "burak@gmail.com",
+                Password = "123456",
+                RefreshToken = "",
+                RefreshTokenExpireDate = DateTime.Now.AddHours(2)
+            };
+            _context.Customers.Add(customer);
+            _context.SaveChanges();
+
+            DeleteCustomerCommand command = new DeleteCustomerCommand(_context, customer.Email, customer.Password);
+            command.Model = new DeleteCustomerModel()
+            {
+                Email = "burak@gmail.com",
+                Password = "123456"
+            };
```
varolmayan bir customer bilgisi vermen gerekiyorken, varolan vermişsin. yukarıda hiç person ve customer oluşturmadan bu test'te varolmayan bir customer epostası vererek bu testi yapman lazım.

fatiyapici replied:
Düzeltildi.

bataylan replied:
Düzelmemiş. Yukarıda person ve customer oluşturmana gerek yok kullanmayacaksan, sadece eposta adresini değiştirmişsin.

bataylan commented on Tests/WebApi.UnitTests/Applications/CustomerOperations/Commands/DeleteCustomer/DeleteCustomerCommandValidatorTest.cs:
```diff
@@ -0,0 +1,44 @@
+            command.Model = new DeleteCustomerModel()
+            {
+                Email = email,
+                Password = password,
```
virgül silelim.

fatiyapici replied:
Düzeltildi.

bataylan commented on Tests/WebApi.UnitTests/Applications/CustomerOperations/Commands/UpdateCustomer/UpdateCustomerCommandTest.cs:
```diff
@@ -0,0 +1,48 @@
+        [Fact]
+        public void WhenUpdatedCustomerIsNotExist_InvalidOperationException_ShouldReturn()
+        {
+            var customer = new Customer()
+            {
+                Name = "Fatih",
+                Surname = "Yapici",
+                Email = "fati@gmail.com",
+                Password = "123456",
+                RefreshToken = ""
+            };
+
+            _context.Customers.Add(customer);
+            _context.SaveChanges();
+
+            UpdateCustomerCommand command = new UpdateCustomerCommand(_context, customer.Id);
+
+            command.Model = new UpdateCustomerViewModel()
+            {
+                Email = "fati@gmail.com",
+                Password = "123456"
+            };
```
test update edilecek customer bulunamaması durumu için. Customer oluşturmaman ve olmayan bir customer eposta adresi vermen gerekiyor.

fatiyapici replied:
Düzetildi fakat test her iki durumda da fail olarak dönüyor. Tekrar kontrol edilecek.

bataylan commented on MovieStoreWebApi/Applications/CustomerOperations/Commands/UpdateCustomer/UpdateCustomerCommand.cs:
```diff
@@ -6,21 +6,26 @@ public class UpdateCustomerCommand
-                throw new InvalidOperationException("Guncellenecek musteri maili yanlis.");
+                throw new InvalidOperationException(ExceptionMessageEmail);
             if (customer.Password != Model.Password)
-                throw new InvalidOperationException("Guncellenecek musteri sifresi yanlis.");
+                throw new InvalidOperationException(ExceptionMessagePassword);
 
             // For Update Mail: customer.Email = Model.Email != default ? Model.Email : customer.Email;
             customer.Password = Model.Password != default ? Model.Password : customer.Password;
```
burada mantık hatası olmuş, Modeldeki şifre ile customer'ı doğruluyorsun ancak sonrasında aynı şifreyle güncelleme yapıyorsun. Şifre customer'ın şifresine eşit değilse de hata veriyorsun, yani aslında hiçbir durumda güncelleme yapmıyorsun.
Model'e NewPassword diye field ekleyerek bu classla ilgili tüm test ve validatorları güncellemen lazım.

Ayrıca `Model.Password != default` sorgusunu buradan çıkarıp validatorına eklemen lazım, ama bu IsEmpty kontrolüne eşit, bunu kullanmanın pek manası yok.

fatiyapici replied:
Düzeltildi.

bataylan commented on Tests/WebApi.UnitTests/Applications/CustomerOperations/Commands/UpdateCustomer/UpdateCustomerCommandValidatorTest.cs:
```diff
@@ -0,0 +1,45 @@
+        [Theory]
+        public void WhenInvalidInputAreGiven_Validator_ShouldBeReturnErrors(string email, string password)
+        {
+            IMovieStoreDbContext dbContext = _context;
```
_context fieldını yeniden değişkene atamanın hiç manası yok, chatgpt ye selamlar.

fatiyapici replied:
Düzeltildi.

bataylan commented on Tests/WebApi.UnitTests/Applications/CustomerOperations/Commands/UpdateCustomer/UpdateCustomerCommandValidatorTest.cs:
```diff
@@ -0,0 +1,45 @@
+        {
+            IMovieStoreDbContext dbContext = _context;
+            int customerId = 1;
+            UpdateCustomerCommand command = new UpdateCustomerCommand(dbContext, customerId);
```
command'e diğer validator testtekiler gibi null değerler verebilirsin.
Id'yi vermek için command'ine field ekle, bu methoda parametre olarak ekleme, diğer testlerinle aynı kalıpta olması için söylüyorum.

fatiyapici replied:
Diğer Validator testleri gibi null değer veremiyorum. `UpdateCustomerCommand` içerisinde bulunanan `Id`'yi kasteediyorsan, yine yapamadım.

bataylan commented on Tests/WebApi.UnitTests/Applications/CustomerOperations/Commands/UpdateCustomer/UpdateCustomerCommandValidatorTest.cs:
```diff
@@ -0,0 +1,45 @@
+            int customerId = 1;
+            UpdateCustomerCommand command = new UpdateCustomerCommand(dbContext, customerId);
+
+            UpdateCustomerViewModel model = new UpdateCustomerViewModel();
+            model.Email = email;
+            model.Password = password;
+            command.Model = model;
```
Bu kısımlar command'le birlikte güncellenecek.

fatiyapici replied:
Anlaşılmadı.

bataylan commented on Tests/WebApi.UnitTests/Applications/CustomerOperations/CreateToken/CreateTokenCommandTest.cs:
```diff
@@ -0,0 +1,50 @@
+    public class CreateTokenCommand
+    {
+        public CreateTokenModel Model { get; set; }
+        public const string ExceptionMessage = "Kullanici adi-sifre hatali.";
```
const field üste alalım

fatiyapici replied:
Düzeltildi.

bataylan commented on Tests/WebApi.UnitTests/Applications/CustomerOperations/CreateToken/CreateTokenCommandTest.cs:
```diff
@@ -0,0 +1,50 @@
+
+        public Token Handle()
+        {
+            var user = _dbContext.Customers.FirstOrDefault(x => x.Email == Model.Email && x.Password == Model.Password);
```
değişken ismi user değil customer olsun

fatiyapici replied:
Düzeltildi.

bataylan commented on Tests/WebApi.UnitTests/Applications/CustomerOperations/CreateToken/CreateTokenCommandTest.cs:
```diff
@@ -0,0 +1,50 @@
+        public Token Handle()
+        {
+            var user = _dbContext.Customers.FirstOrDefault(x => x.Email == Model.Email && x.Password == Model.Password);
+            if (user is not null)
+            {
+                MovieStoreWebApi.TokenOperations.Models.TokenHandler handler = new MovieStoreWebApi.TokenOperations.Models.TokenHandler(_configuration);
+                Token token = handler.CreateAccessToken(user);
+
+                user.RefreshToken = token.RefreshToken;
+                user.RefreshTokenExpireDate = token.Expiration.AddMinutes(5);
+                _dbContext.SaveChanges();
+                
+                return token;
+            }
+            else
+            {
+                throw new InvalidOperationException(ExceptionMessage);
+            }
```
if içindeki kodu dışarı çıkar, if kontrolünü de is null ise yapıp orada exception ver, öylesi daha iyi olur.

MovieStoreWebApi.TokenOperations.Models bunu bu şekilde kullanmak yerine using ile ekleyelim.

fatiyapici replied:
Düzeltildi.

bataylan reviewed (changes requested):
Biraz kafan karışmış gibi duruyor. Birkaç command ve testte mantık hataları var. Diğer testlerle birebir olmadığı için Customer tarafı biraz zorlamış gibi. O kısımları güncellediğinde tekrar kontrol edelim.
Bunun dışında field sıralaması ve indetation vardı yine, çok önemli değil.

fatiyapici committed:
update create customer unit test const fix

fatiyapici committed:
update delete customer unit test name and const fix

fatiyapici committed:
update get customer detail unit test identation

fatiyapici committed:
update create token unit test condition revert-const

fatiyapici committed:
update update customer unit test

UpdateCustomerCommandValidatorTest içerisinde hata mevcut incelenip, tekrar bakılacak.

fatiyapici commented:
Mantıksal hatalar, identation ve const hataları düzeltildi.
`UpdateCustomerCommand` içerisinde şifre güncellemesi tekrar kontrol edilse iyi olur.
`UpdateCustomerCommandValidatorTest` hatalı.

bataylan commented on MovieStoreWebApi/Applications/CustomerOperations/Commands/DeleteCustomer/DeleteCustomerCommand.cs:
```diff
@@ -4,38 +4,36 @@ namespace MovieStoreWebApi.Applications.CustomerOperations.DeleteCustomer
 
         public DeleteCustomerCommand(IMovieStoreDbContext dbContext, string email, string password)
         {
             _dbContext = dbContext;
-            email = CustomerEmail;
-            password = CustomerPassword;
+            email = Model.Email;
+            password = Model.Password;
```
Bu yapılmayacak bir hata, bunu baktığında görebiliyor olman lazım

Email ve password zaten argüman olarak veriliyor bu metodun kullanıldığı yerden bu metodun sonunda da bu değişkenler yok oluyor zaten, classta başka bir yerde kullanılmıyor. Bu argümanları modeldeki değerle güncellemenin nasıl bir mantığı var?

fatiyapici replied:
email ve password silindi.

bataylan commented on MovieStoreWebApi/Applications/CustomerOperations/Commands/UpdateCustomer/UpdateCustomerCommand.cs:
```diff
@@ -4,11 +4,11 @@ namespace MovieStoreWebApi.Applications.CustomerOperations.UpdateCustomer
         public const string ExceptionMessageFound = "Guncellenecek musteri bulunamadi.";
         public const string ExceptionMessageEmail = "Guncellenecek musteri maili yanlis.";
         public const string ExceptionMessagePassword = "Guncellenecek musteri sifresi yanlis.";
+        public UpdateCustomerViewModel Model { get; set; }
```
boşluk lütfen

bataylan commented on Tests/WebApi.UnitTests/Applications/CustomerOperations/Commands/DeleteCustomer/DeleteCustomerCommandTest.cs:
```diff
@@ -45,7 +45,7 @@ public void WhenDeletedCustomerIsNotExist_InvalidOperationException_ShouldReturn
-                Email = "burak@gmail.com",
+                Email = "hassemercioglu@gmail.com",
                 Password = "123456"
             };
```
Yukarıda person ve customer niye oluşturuyoruz? O kısımları niye kaldırmadın (26-43) ?

fatiyapici replied:
Kaldırıldı.

bataylan commented on Tests/WebApi.UnitTests/Applications/CustomerOperations/Commands/UpdateCustomer/UpdateCustomerCommandTest.cs:
```diff
@@ -36,7 +36,7 @@ public void WhenUpdatedCustomerIsNotExist_InvalidOperationException_ShouldReturn
-                Email = "fati@gmail.com",
+                Email = "yapici@gmail.com",
                 Password = "123456"
             };
```
Düzeltilmemiş. Kontrol edilecek demişsin ama hatanın sebebi muhtemelen zaten bu söyleyeceğim şey. Customer'ın bulunmaması isteniyor zaten, diğer testte de aynısını yapmışsın, madem kullanmayacaksak yukarıda customer oluşturmana gerek yok.

Yani yukarıda customer oluşturduğun kısmı neden silmiyorsun (23-33). Ayrıca bunun command'i yanlıştı zaten diye hatırlıyorum.

bataylan commented on MovieStoreWebApi/Applications/CustomerOperations/Commands/UpdateCustomer/UpdateCustomerCommand.cs:
```diff
@@ -28,7 +28,7 @@ public void Handle()
 
             // For Update Mail: customer.Email = Model.Email != default ? Model.Email : customer.Email;
-            customer.Password = Model.Password != default ? Model.Password : customer.Password;
+            customer.Password = customer.Password != Model.Password ? Model.NewPassword : customer.Password;
```
Burada konuştuğumuz mantık hatası giderilmemiş, lütfen üstteki satırları da okuyarak güncelle burayı.

fatiyapici replied:
Güncellendi.

bataylan reviewed (changes requested):
Bu değişiklikler olmamış, özellikle mantık hataları olduğu gibi duruyor, düzeltilmesi gereken asıl şeyler gidip eposta adresi değiştirilerek düzeltilmiş. Daha özenli bir güncelleme yap, hataları da gidericem derken ide hata göstermeden mantıksız bir şey yazarak geçmemek gerekiyor.

fatiyapici committed:
fix delete customer unit test

fatiyapici committed:
fix update customer unit test new password

fatiyapici committed:
update get customer unit test const lines

fatiyapici committed:
update create token unit test const line

fatiyapici commented:
`UpdateCustomer` içerisindeki `NewPassword` sorunu çözüldü.
const satırları tekrar düzenlendi.

bataylan commented on Tests/WebApi.UnitTests/Applications/CustomerOperations/Commands/UpdateCustomer/UpdateCustomerCommandTest.cs:
```diff
@@ -0,0 +1,48 @@
+        [Fact]
+        public void WhenUpdatedCustomerIsNotExist_InvalidOperationException_ShouldReturn()
+        {
+            var customer = new Customer()
+            {
+                Name = "Fatih",
+                Surname = "Yapici",
+                Email = "fati@gmail.com",
+                Password = "123456",
+                RefreshToken = ""
+            };
+
+            _context.Customers.Add(customer);
+            _context.SaveChanges();
+
+            UpdateCustomerCommand command = new UpdateCustomerCommand(_context, customer.Id);
```
Bu test bulunamayan bir customer icin gozukuyor ama sen customer'i olusturup id'sini vermissin, dolayisiyla customeri her zaman bulmayacak mi?

fatiyapici replied:
Daha önce oluşturulan customer silindi ve command'e manuel id verilerek düzeltildi.

bataylan reviewed (changes requested):
Bir test yanlis gozukuyor, digerlerinde yazima bakmistim daha cok umarim atlamamisimdir, onlari da kontrol et istersen ayni mantik hatasi var mi diye.

fatiyapici committed:
fix update customer unit test

fatiyapici commented:
UpdateCustomerCommandTest içerisindeki case düzeltildi.

bataylan reviewed (approved):
Sorunlar cozulmus gorunuyor, eline saglik 👍
